feat(unmarshal): introduce new unmarshal functions for primitives and models

padamstx opened the pull request:
Fixes: https://github.ibm.com/arf/planning-sdk-squad/issues/1613

This PR contains new unmarshal functions for primitive and model types in the Go core.
This will be needed in the SDK generator changes to fix the referenced issue.

The "new" approach for unmarshalling JSON values (primitives and models) in the Go generator will be to generate a single `Unmarshal<model>()` function for each model.   This generated function can then be passed as a parameter (along with the JSON input and a pointer to the unmarshal result) to the core's UnmarshalModel() function.  UnmarshalModel() will use reflection to interpret the JSON input and the result to determine the appropriate type of unmarshal operation (e.g. a single model instance, a slice of models, a map of models, or a map of model slices).

The new `UnmarshalPrimitive()` function will be invoked from within a model's generated `Unmarshal<model>()` function to unmarshal struct fields involving primitive types (e.g. a single instance of a primitive type, a slice of primitives, a map of primitives, a map of primitive slices, and a slice of maps of primitive slices).

This PR contains only the changes in the Go core.  There will be a corresponding PR for the SDK generator changes.

Note: there might be some additional tweaks needed to the code changes in this PR as I complete the corresponding SDK generator changes, but I wanted to get this PR submitted so the review process can begin.   One change that I identified as I reviewed the code before submitting this PR is that there is a common piece of code within each of the private unmarshalModelXXX functions that I can pull out and implement in a private function.   This will not change the overall behavior of those functions.

padamstx commented:
@smikulcik Hi Simon, I wanted to tag you as a reviewer on this PR, but I see you're not in the required team within the github.com/IBM org.    Please feel free to take a look and provide feedback one way or another :)

jorge-ibm commented on core/unmarshal_v2.go:
```diff
@@ -0,0 +1,521 @@
+				return
+			}
+		} else {
+			rawMap = nil
```
if ``rawMap`` is nil this method will return no error, but what is the expected unmarshal result? Will ``result`` still be an empty interface?

padamstx replied:
In the case where the specified property is not found in the input map, we want to essentially return a nil model pointer.
That's why we initialize te result to nil, then do the actual unmarshalling only if rawMap is not nil.

jorge-ibm commented on core/unmarshal_v2.go:
```diff
@@ -0,0 +1,521 @@
+			}
+		}
+	} else {
+		foundIt = true
```
Suggestion: set ``foundIt`` to ``true`` when you first initialize it so you can remove this extra condition. If ``rawMap`` doesn't have the property then it'll be false, otherwise it'll still be set to true

mkistler commented on core/unmarshal_v2_primitives_test.go:
```diff
@@ -0,0 +1,1386 @@
+
+	// Negative tests
+	model.Prop = nil
+	err = UnmarshalPrimitive(rawMap, "bad_type", &model.Prop)
```
This seems wrong to me.  Type "Any" should allow any type -- how can we have a bad type?

padamstx replied:
Come on Mike, I was using alliteration in my testcase development :)  lol

mkistler replied:
hahahaha

padamstx commented on core/unmarshal_v2_primitives_test.go:
```diff
@@ -0,0 +1,1386 @@
+		"bad_type":  true,
+		"not_a_slice": false,
+		"bad_slice_type": [38, 26],
+		"null_prop": null
```
It turns out that "" and "not a real uuid" are not treated as errors by the strfmt.UUID unmarshalling code.   Not surprising due to the fact that UUID is really just an alias for "string":
```
// UUID represents a uuid string format
//
// swagger:strfmt uuid
type UUID string
```

padamstx commented on core/unmarshal_v2_primitives_test.go:
```diff
@@ -0,0 +1,1386 @@
+		"bad_type":  true,
+		"not_a_slice": false,
+		"bad_slice_type": [38, 26],
+		"null_prop": null
```
it turns out that "" will be unmarshalled by the strfmt.DateTime unmarshalling code to produce "1970-01-01T00:00:00.000Z", rather than produce an error.

mkistler replied:
Ouch!

padamstx commented:
I just pushed a new commit to this PR that includes the following:
- Addresses review comments
- Adds a bunch of additional tests, particularly negative tests
- While adding more tests, I realized that I needed to add tests involving models with a descriminator, and that required some minor changes to the existing function to accomodate both `**Foo` and `*FooIntf` as model result types that need to be supported.

I think (I hope) that this PR is now 100% complete and ready for the corresponding Generator changes to commence (famous last words, right?).

padamstx commented:
While working through the Go generator changes, I found even more opportunities to improve the Go core :)
I just pushed a commit that enhances the BaseService.Request() method so that it can directly return various types of response body data via the 'result' parameter (and the DetailedResponse.Result field as well).   These changes allow us to reduce the amount of code that we generate immediately after the call to BaseService.Request() within an operation method.

ibm-devx-automation commented:
:tada: This PR is included in version 3.4.0 :tada:

The release is available on [GitHub release](https://github.com/IBM/go-sdk-core/releases/tag/v3.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

ibm-devx-automation commented:
:tada: This PR is included in version 4.0.0 :tada:

The release is available on [GitHub release](https://github.com/IBM/go-sdk-core/releases/tag/v4.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
